Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp

```diff
@@ -0,0 +1,150 @@
+#include <EditorPluginAssets/EditorPluginAssetsPCH.h>
+
+#include <EditorPluginVisualScript/VisualScriptGraph/VisualScriptGraph.h>
+#include <EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.h>
+
+// static
+ezVisualScriptDataType::Enum ezVisualScriptTypeDeduction::DeductFromNodeDataType(const ezVisualScriptPin& pin)
+{
+  auto pObject = pin.GetParent();
+  auto pManager = static_cast<const ezVisualScriptNodeManager*>(pObject->GetDocumentObjectManager());
+
+  return pManager->GetDeductedType(pObject);
+}
+
+// static
+ezVisualScriptDataType::Enum ezVisualScriptTypeDeduction::DeductFromTypeProperty(const ezVisualScriptPin& pin)
+{
+  if (auto pType = GetReflectedType(pin.GetParent()))
+  {
+    return ezVisualScriptDataType::FromRtti(pType);
+  }
+
+  return ezVisualScriptDataType::Invalid;
+}
+
+// static
+ezVisualScriptDataType::Enum ezVisualScriptTypeDeduction::DeductFromAllInputPins(const ezDocumentObject* pObject, const ezVisualScriptPin* pDisconnectedPin)
+{
+  auto pManager = static_cast<const ezVisualScriptNodeManager*>(pObject->GetDocumentObjectManager());
+
+  ezVisualScriptDataType::Enum deductedType = ezVisualScriptDataType::Invalid;
+
+  ezHybridArray<const ezVisualScriptPin*, 16> pins;
+  pManager->GetInputDataPins(pObject, pins);
+  for (auto pPin : pins)
+  {
+    if (pPin->GetScriptDataType() != ezVisualScriptDataType::Any)
+      continue;
+
+    // the pin is about to be disconnected so we ignore it here
+    if (pPin == pDisconnectedPin)
+      continue;
+
+    ezVisualScriptDataType::Enum pinDataType = ezVisualScriptDataType::Invalid;
+    auto connections = pManager->GetConnections(*pPin);
+    if (connections.IsEmpty() == false)
+    {
+      pinDataType = static_cast<const ezVisualScriptPin&>(connections[0]->GetSourcePin()).GetResolvedScriptDataType();
+    }
+    else
+    {
+      ezVariant var = pObject->GetTypeAccessor().GetValue(pPin->GetName());
+      pinDataType = ezVisualScriptDataType::FromVariantType(var.GetType());
+    }
+
+    deductedType = ezMath::Max(deductedType, pinDataType);
```

Comment:
Is it commented at ezVisualScriptDataType that the order is important ?
Maybe even add some static_asserts to prevent future changes.

Reply:
Added a comment